### PR TITLE
Adding support for close and clear on shutdown

### PR DESCRIPTION
**Pull request by sandvige**
This pull request attempts to handle this case: https://github.com/doctrine/DoctrineBundle/issues/638

### CODE REVIEW

File: DoctrineBundle.php [hunk 1]
```diff
@@ -119,17 +119,21 @@ public function shutdown()
         // Clear all entity managers to clear references to entities for GC
         if ($this->container->hasParameter('doctrine.entity_managers')) {
             foreach ($this->container->getParameter('doctrine.entity_managers') as $id) {
-                if (!$this->container instanceof IntrospectableContainerInterface || $this->container->initialized($id)) {
-                    $this->container->get($id)->clear();
+                if ($this->container->getParameter('doctrine.orm.entity_managers.' . $id . '.clear_on_shutdown') === true) {
```
Comment (Ocramius) [inline DoctrineBundle.php]: Can you split this to private methods?

File: DoctrineBundle.php [hunk 2]
```diff
@@ -119,17 +119,21 @@ public function shutdown()
...
                 }
             }
         }
 
         // Close all connections to avoid reaching too many connections in the process when booting again later (tests)
         if ($this->container->hasParameter('doctrine.connections')) {
             foreach ($this->container->getParameter('doctrine.connections') as $id) {
-                if (!$this->container instanceof IntrospectableContainerInterface || $this->container->initialized($id)) {
-                    $this->container->get($id)->close();
+                if ($this->container->getParameter('doctrine.dbal.connections.' . $id . '.close_on_shutdown') === true) {
```
Comment (Ocramius) [inline DoctrineBundle.php]: Same as above

File: DoctrineBundle.php (hunk 1 above)
Comment (stof) [inline DoctrineBundle.php]: Putting the whole service id inside the parameter id looks wrong to me.

And your never use the configuration to configure this parameter, so it won't work.

Comment (sandvige) [inline DoctrineBundle.php]: Where do you prefer privates methods? At start or end of the file?

**Comment by sandvige** [comment]
@Ocramius @stof We fixed unit tests, but it looks like there's a problem with deprecated calls. Should we investigate the error or this is already done into another PR that is handling it?
```
Resetting a non-lazy manager service is deprecated since Symfony 3.2 and will throw an exception in version 4.0. Set the "doctrine.orm.default_entity_manager" service as lazy and require "symfony/proxy-manager-bridge" in your composer.json file instead
```

**Comment by sandvige** [comment]
Ok everything is fixed after a merge with upstream

**Comment by sandvige** [comment]
@Ocramius ? @stof ?

**Comment by kimhemsoe** [comment]
Is this ready for merge?

**Comment by mikeSimonson** [comment]
@kimhemsoe @sandvige I have the impression that there is not test to verify that close_on_shutdown is called when it's set and not called when it's not. Same for clear_on_shutdown .

**Comment by kimhemsoe** [comment]
ping @sandvige

**Comment by kimhemsoe** [comment]
@sandvige Can you provide the missing tests?

**Comment by kimhemsoe** [comment]
@sandvige Sorry i am closing this for now. If you or someone else have time in the future to add the missing items i will ofc reopen.

**Pull request closed without merging**